rmw_dds_common/include/rmw_dds_common/context.hpp

```diff
@@ -49,6 +49,151 @@ struct Context
+  RMW_DDS_COMMON_PUBLIC
+  rmw_ret_t
+  update_publisher_graph(
+    const rmw_gid_t & publisher_gid, const std::string & name, const std::string & namespace_);
+
+  /// Update graph for destroying a publisher.
+  /**
+   * \param publisher_gid publisher gid.
+   * \param name node name.
+   * \param namespace_ node namespace.
+   * \return `RMW_RET_OK` if successful, or
+   * \return `RMW_RET_ERROR` an unexpected error occurs.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  rmw_ret_t
+  destroy_publisher_graph(
+    const rmw_gid_t & publisher_gid, const std::string & name, const std::string & namespace_);
+
+  /// Update graph for creating a client.
+  /**
+   * \param request_publisher_gid request publisher gid of the client.
+   * \param response_subscriber_gid response subscriber gid of the client.
+   * \param name node name.
+   * \param namespace_ node namespace.
+   * \return `RMW_RET_OK` if successful, or
+   * \return `RMW_RET_ERROR` an unexpected error occurs.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  rmw_ret_t
+  update_client_graph(
+    const rmw_gid_t & request_publisher_gid, const rmw_gid_t & response_subscriber_gid,
+    const std::string & name, const std::string & namespace_);
+
+  /// Update graph for destroying a client.
+  /**
+   * \param request_publisher_gid request publisher gid of the client.
+   * \param response_subscriber_gid response subscriber gid of the client.
+   * \param name node name.
+   * \param namespace_ node namespace.
+   * \return `RMW_RET_OK` if successful, or
+   * \return `RMW_RET_ERROR` an unexpected error occurs.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  rmw_ret_t
+  destroy_client_graph(
+    const rmw_gid_t & request_publisher_gid, const rmw_gid_t & response_subscriber_gid,
+    const std::string & name, const std::string & namespace_);
+
+  /// Update graph for creating a service.
+  /**
+   * \param request_subscriber_gid request subscriber gid of the client.
+   * \param response_publisher_gid response publisher gid of the client.
+   * \param name node name.
+   * \param namespace_ node namespace.
+   * \return `RMW_RET_OK` if successful, or
+   * \return `RMW_RET_ERROR` an unexpected error occurs.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  rmw_ret_t
+  update_service_graph(
+    const rmw_gid_t & request_subscriber_gid, const rmw_gid_t & response_publisher_gid,
+    const std::string & name, const std::string & namespace_);
+
+  /// Update graph for destroying a service.
+  /**
+   * \param request_subscriber_gid request subscriber gid of the client.
+   * \param response_publisher_gid response publisher gid of the client.
+   * \param name node name.
+   * \param namespace_ node namespace.
+   * \return `RMW_RET_OK` if successful, or
+   * \return `RMW_RET_ERROR` an unexpected error occurs.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  rmw_ret_t
+  destroy_service_graph(
+    const rmw_gid_t & request_subscriber_gid, const rmw_gid_t & response_publisher_gid,
+    const std::string & name, const std::string & namespace_);
+
+private:
```

Comment:
Given this, and all of the additional functionality we are adding to this `Context`, should we just make it a class at this point?  That would allow us to (potentially) hide more of the member variables.

But I don't feel strongly about this point, so if you think we should leave this as a `struct` I'm OK with that too.

Reply:
I think it's better to leave them into another refactor PR.